src/constraint.rs

```diff
@@ -1500,3 +1522,61 @@ impl ConstraintValidator for TimestampPrecisionConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `valid_values` constraint
+/// [valid_values]: https://amzn.github.io/ion-schema/docs/spec.html#valid_values
+#[derive(Debug, Clone, PartialEq)]
+pub enum ValidValuesConstraint {
+    Range(Range),
+    Values(Vec<OwnedElement>),
+}
+
+impl ValidValuesConstraint {}
+
+impl ConstraintValidator for ValidValuesConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        return match self {
+            ValidValuesConstraint::Range(range) => match value.ion_type() {
+                IonType::Integer | IonType::Float | IonType::Decimal | IonType::Timestamp => {
+                    if range.contains(value).unwrap() {
+                        Ok(())
+                    } else {
+                        Err(Violation::new(
+                            "valid_values",
+                            ViolationCode::InvalidValue,
+                            &format!(
+                                "expected valid_values to be from {:?} found {:?}",
+                                &self, value
+                            ),
+                        ))
+                    }
+                }
+                _ => Err(Violation::new(
+                    "valid_values",
+                    ViolationCode::TypeMismatched,
+                    &format!(
+                        "expected valid_values to be a range or list but found {:?}",
```

Comment:
Looks like a copy/paste with the wrong message. The value is not valid because it's not in the range.

Reply:
Actually this one checks for `IonType` for the given value if its not a number/timestamp and the `valid_values` is a range then it returns violation or rather this is illegal?

Reply:
Okay, I think it might just be that the phrasing is confusing to me—It sounds like `valid_values` is referring to the constraint and this is a problem with the schema being invalid. Perhaps "expected value to be from {:?} but found {:?}", like the other violation messages for this constraint.